[quant] Implement quantize APoT method

asl3 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #79845
* #79940
* #79530
* __->__ #79499
* #79936

### Summary:
This PR implements functionality to quantize input floating point tensors. The method takes a floating point tensor as input and converts to APoT integer representation based on the calculated quantization levels from the specified APoT non-uniform observer. This approach follows the methodology outlined in the paper https://arxiv.org/pdf/1909.13144.pdf.

### Test Plan:
Run unit tests with: `python pytorch/test/quantization/core/experimental/test_quantizer.py`

asl3 committed:
[quant] Implement quantize APoT method

[ghstack-poisoned]

vadimkantorov commented:
naming suggestion: call it APOT or apot. APoT will be hard to type and easy to to forget which letter is small especially when you need to type it many times...

asl3 committed:
Update on "[quant] Implement quantize APoT method"




[ghstack-poisoned]

asl3 committed:
Update on "[quant] Implement quantize APoT method"




[ghstack-poisoned]

asl3 committed:
Update on "[quant] Implement quantize APoT method"




[ghstack-poisoned]

asl3 committed:
Update on "[quant] Implement quantize APoT method"




[ghstack-poisoned]

asl3 committed:
Update on "[quant] Implement quantize APoT method"




[ghstack-poisoned]

asl3 committed:
Update on "[quant] Implement quantize APoT method"




[ghstack-poisoned]

asl3 committed:
Update on "[quant] Implement quantize APoT method"




[ghstack-poisoned]

asl3 committed:
Update on "[quant] Implement quantize APoT method"




[ghstack-poisoned]

asl3 committed:
Update on "[quant] Implement quantize APoT method"


### Summary:
This PR implements functionality to quantize input floating point tensors. The method takes a floating point tensor as input and converts to APoT integer representation based on the calculated quantization levels from the specified APoT non-uniform observer. This approach follows the methodology outlined in the paper https://arxiv.org/pdf/1909.13144.pdf.

### Test Plan:
Run unit tests with: python pytorch/test/quantization/core/experimental/test_quantized_tensor.py

[ghstack-poisoned]

asl3 committed:
Update on "[quant] Implement quantize APoT method"


### Summary:
This PR implements functionality to quantize input floating point tensors. The method takes a floating point tensor as input and converts to APoT integer representation based on the calculated quantization levels from the specified APoT non-uniform observer. This approach follows the methodology outlined in the paper https://arxiv.org/pdf/1909.13144.pdf.

### Test Plan:
Run unit tests with: `python pytorch/test/quantization/core/experimental/test_quantized_tensor.py`

[ghstack-poisoned]

asl3 committed:
Update on "[quant] Implement quantize APoT method"


### Summary:
This PR implements functionality to quantize input floating point tensors. The method takes a floating point tensor as input and converts to APoT integer representation based on the calculated quantization levels from the specified APoT non-uniform observer. This approach follows the methodology outlined in the paper https://arxiv.org/pdf/1909.13144.pdf.

### Test Plan:
Run unit tests with: `python pytorch/test/quantization/core/experimental/test_quantized_tensor.py`

[ghstack-poisoned]

asl3 committed:
Update on "[quant] Implement quantize APoT method"


### Summary:
This PR implements functionality to quantize input floating point tensors. The method takes a floating point tensor as input and converts to APoT integer representation based on the calculated quantization levels from the specified APoT non-uniform observer. This approach follows the methodology outlined in the paper https://arxiv.org/pdf/1909.13144.pdf.

### Test Plan:
Run unit tests with: `python pytorch/test/quantization/core/experimental/test_quantized_tensor.py`

[ghstack-poisoned]

asl3 committed:
Update on "[quant] Implement quantize APoT method"


### Summary:
This PR implements functionality to quantize input floating point tensors. The method takes a floating point tensor as input and converts to APoT integer representation based on the calculated quantization levels from the specified APoT non-uniform observer. This approach follows the methodology outlined in the paper https://arxiv.org/pdf/1909.13144.pdf.

### Test Plan:
Run unit tests with: `python pytorch/test/quantization/core/experimental/test_quantized_tensor.py`

[ghstack-poisoned]

asl3 committed:
Update on "[quant] Implement quantize APoT method"


### Summary:
This PR implements functionality to quantize input floating point tensors. The method takes a floating point tensor as input and converts to APoT integer representation based on the calculated quantization levels from the specified APoT non-uniform observer. This approach follows the methodology outlined in the paper https://arxiv.org/pdf/1909.13144.pdf.

### Test Plan:
Run unit tests with: `python pytorch/test/quantization/core/experimental/test_quantized_tensor.py`

[ghstack-poisoned]

asl3 committed:
Update on "[quant] Implement quantize APoT method"


### Summary:
This PR implements functionality to quantize input floating point tensors. The method takes a floating point tensor as input and converts to APoT integer representation based on the calculated quantization levels from the specified APoT non-uniform observer. This approach follows the methodology outlined in the paper https://arxiv.org/pdf/1909.13144.pdf.

### Test Plan:
Run unit tests with: `python pytorch/test/quantization/core/experimental/test_quantized_tensor.py`

[ghstack-poisoned]

asl3 committed:
Update on "[quant] Implement quantize APoT method"


### Summary:
This PR implements functionality to quantize input floating point tensors. The method takes a floating point tensor as input and converts to APoT integer representation based on the calculated quantization levels from the specified APoT non-uniform observer. This approach follows the methodology outlined in the paper https://arxiv.org/pdf/1909.13144.pdf.

### Test Plan:
Run unit tests with: `python pytorch/test/quantization/core/experimental/test_quantized_tensor.py`

[ghstack-poisoned]

asl3 committed:
Update on "[quant] Implement quantize APoT method"


### Summary:
This PR implements functionality to quantize input floating point tensors. The method takes a floating point tensor as input and converts to APoT integer representation based on the calculated quantization levels from the specified APoT non-uniform observer. This approach follows the methodology outlined in the paper https://arxiv.org/pdf/1909.13144.pdf.

### Test Plan:
Run unit tests with: `python pytorch/test/quantization/core/experimental/test_quantized_tensor.py`

[ghstack-poisoned]

asl3 committed:
Update on "[quant] Implement quantize APoT method"


### Summary:
This PR implements functionality to quantize input floating point tensors. The method takes a floating point tensor as input and converts to APoT integer representation based on the calculated quantization levels from the specified APoT non-uniform observer. This approach follows the methodology outlined in the paper https://arxiv.org/pdf/1909.13144.pdf.

### Test Plan:
Run unit tests with: `python pytorch/test/quantization/core/experimental/test_quantized_tensor.py`

[ghstack-poisoned]

asl3 committed:
Update on "[quant] Implement quantize APoT method"


### Summary:
This PR implements functionality to quantize input floating point tensors. The method takes a floating point tensor as input and converts to APoT integer representation based on the calculated quantization levels from the specified APoT non-uniform observer. This approach follows the methodology outlined in the paper https://arxiv.org/pdf/1909.13144.pdf.

### Test Plan:
Run unit tests with: `python pytorch/test/quantization/core/experimental/test_quantized_tensor.py`

[ghstack-poisoned]

dzdang commented on test/quantization/core/experimental/test_quantized_tensor.py:
```diff
@@ -1,19 +1,116 @@
+        size = random.randint(1, 16)
+
+        # generate tensor with random fp values between 0 -> 1
+        tensor2quantize = torch.rand(size)
```
2 questions
1) why does `tensor2quantize`'s size have to be equal to the number of quantization levels (in a comment above, you seem to imply this)?

2) we don't need to restrict fp values in [0, 1]. it can be unbounded on both sides

asl3 replied:
agreed, will pick arbitrary dimension sizes

dzdang commented on test/quantization/core/experimental/test_quantized_tensor.py:
```diff
@@ -1,19 +1,116 @@
+        # generate tensor with random fp values between 0 -> 1
+        tensor2quantize = torch.rand(size)
+
+        apot_tens = TensorAPoT(4, 1, False)
```
tens = tensor? 
better to spell it out the extra 2 letters because "tens" meaning is ambiguous. alternatively, we can just call it qtensor (a variable name used a lot in the quantization test suite)

dzdang commented on test/quantization/core/experimental/test_quantized_tensor.py:
```diff
@@ -1,19 +1,116 @@
+        # get apot quantized tensor result
+        apot_tens = apot_tens.quantize_APoT(tensor2quantize=tensor2quantize, apot_repr=APoTRepr.level_indices)
+
+        # get uniform observer quantized tensor result
```
observer?

dzdang replied:
here too

dzdang commented on test/quantization/core/experimental/test_quantized_tensor.py:
```diff
@@ -1,19 +1,116 @@
+    def test_quantize_APoT_rand_1d(self):
+        # generate random size of tensor2dequantize between 1 -> 16
+        # because there are 2**b = 2**4 quantization levels total
+        size = random.randint(1, 16)
```
we can use torch.randint

asl3 replied:
i'm using torch.rand to generate the tensors so i can get fp values, random.randint was to generate random sizes for the dimensions

dzdang commented on test/quantization/core/experimental/test_quantized_tensor.py:
```diff
@@ -1,19 +1,116 @@
+                expectedResult = False
+
+        self.assertTrue(expectedResult)
 
```
these test cases only use k=1, so currently passing the test cases only tells us that uniform quantization is indeed a special case of APoT, but what about non-uniform cases of APoT? 
 we should have some test cases for those as well to be comprehensive

asl3 replied:
@dzdang I have test cases for other values of k in my dequantize PR tests, since I test by calling dequant -> quant. Could I make a note of that in this PR?

dzdang replied:
this PR should stand on its own with the appropriate test cases. it shouldn't have a dependency on dequant

dzdang commented on torch/ao/quantization/experimental/APoT_tensor.py:
```diff
@@ -1,13 +1,66 @@
+# enum to represent APoT representation
+class APoTRepr(enum.Enum):
+    level_indices = 1
+    reduced_precision_fp = 2
```
since there are only 2 known representations of APoT, we could just have a boolean flag instead of an enum. maybe 
```
use_int_repr
```

dzdang commented on torch/ao/quantization/experimental/APoT_tensor.py:
```diff
@@ -1,13 +1,66 @@
+        self.apot_repr = apot_repr
+        if apot_repr == APoTRepr.level_indices:
+            # map float_to_apot over tensor2quantize elements
+            self.data = tensor2quantize.apply_(lambda x: float_to_apot(x, self.quantization_levels, self.level_indices))
```
might be better to refactor some of these functions to a utility file in a separate PR isntead of having it in observer.py and is not a property of the observer. `float_to_apot` is used by your observer visualization and when quantizing a tensor.

dzdang commented on torch/ao/quantization/experimental/APoT_tensor.py:
```diff
@@ -1,13 +1,66 @@
+            self.data = tensor2quantize.apply_(lambda x: float_to_apot(x, self.quantization_levels, self.level_indices))
+        elif apot_repr == APoTRepr.reduced_precision_fp:
+            self.data = tensor2quantize.apply_(lambda x:
+                                               float_to_reduced_precision(x, self.quantization_levels, self.level_indices))
```
same here

asl3 committed:
Update on "[quant] Implement quantize APoT method"


### Summary:
This PR implements functionality to quantize input floating point tensors. The method takes a floating point tensor as input and converts to APoT integer representation based on the calculated quantization levels from the specified APoT non-uniform observer. This approach follows the methodology outlined in the paper https://arxiv.org/pdf/1909.13144.pdf.

### Test Plan:
Run unit tests with: `python pytorch/test/quantization/core/experimental/test_quantized_tensor.py`

[ghstack-poisoned]

asl3 committed:
Update on "[quant] Implement quantize APoT method"


### Summary:
This PR implements functionality to quantize input floating point tensors. The method takes a floating point tensor as input and converts to APoT integer representation based on the calculated quantization levels from the specified APoT non-uniform observer. This approach follows the methodology outlined in the paper https://arxiv.org/pdf/1909.13144.pdf.

### Test Plan:
Run unit tests with: `python pytorch/test/quantization/core/experimental/test_quantized_tensor.py`

[ghstack-poisoned]

asl3 committed:
Update on "[quant] Implement quantize APoT method"


### Summary:
This PR implements functionality to quantize input floating point tensors. The method takes a floating point tensor as input and converts to APoT integer representation based on the calculated quantization levels from the specified APoT non-uniform observer. This approach follows the methodology outlined in the paper https://arxiv.org/pdf/1909.13144.pdf.

### Test Plan:
Run unit tests with: `python pytorch/test/quantization/core/experimental/test_quantized_tensor.py`

[ghstack-poisoned]

asl3 committed:
Update on "[quant] Implement quantize APoT method"


### Summary:
This PR implements functionality to quantize input floating point tensors. The method takes a floating point tensor as input and converts to APoT integer representation based on the calculated quantization levels from the specified APoT non-uniform observer. This approach follows the methodology outlined in the paper https://arxiv.org/pdf/1909.13144.pdf.

### Test Plan:
Run unit tests with: `python pytorch/test/quantization/core/experimental/test_quantized_tensor.py`

[ghstack-poisoned]

asl3 committed:
Update on "[quant] Implement quantize APoT method"


### Summary:
This PR implements functionality to quantize input floating point tensors. The method takes a floating point tensor as input and converts to APoT integer representation based on the calculated quantization levels from the specified APoT non-uniform observer. This approach follows the methodology outlined in the paper https://arxiv.org/pdf/1909.13144.pdf.

### Test Plan:
Run unit tests with: `python pytorch/test/quantization/core/experimental/test_quantized_tensor.py`

[ghstack-poisoned]

asl3 committed:
Update on "[quant] Implement quantize APoT method"


### Summary:
This PR implements functionality to quantize input floating point tensors. The method takes a floating point tensor as input and converts to APoT integer representation based on the calculated quantization levels from the specified APoT non-uniform observer. This approach follows the methodology outlined in the paper https://arxiv.org/pdf/1909.13144.pdf.

### Test Plan:
Run unit tests with: `python pytorch/test/quantization/core/experimental/test_quantized_tensor.py`

[ghstack-poisoned]

asl3 committed:
Update on "[quant] Implement quantize APoT method"


### Summary:
This PR implements functionality to quantize input floating point tensors. The method takes a floating point tensor as input and converts to APoT integer representation based on the calculated quantization levels from the specified APoT non-uniform observer. This approach follows the methodology outlined in the paper https://arxiv.org/pdf/1909.13144.pdf.

### Test Plan:
Run unit tests with: `python pytorch/test/quantization/core/experimental/test_quantized_tensor.py`

[ghstack-poisoned]

asl3 committed:
Update on "[quant] Implement quantize APoT method"


### Summary:
This PR implements functionality to quantize input floating point tensors. The method takes a floating point tensor as input and converts to APoT integer representation based on the calculated quantization levels from the specified APoT non-uniform observer. This approach follows the methodology outlined in the paper https://arxiv.org/pdf/1909.13144.pdf.

### Test Plan:
Run unit tests with: `python pytorch/test/quantization/core/experimental/test_quantized_tensor.py`

[ghstack-poisoned]

asl3 committed:
Update on "[quant] Implement quantize APoT method"


### Summary:
This PR implements functionality to quantize input floating point tensors. The method takes a floating point tensor as input and converts to APoT integer representation based on the calculated quantization levels from the specified APoT non-uniform observer. This approach follows the methodology outlined in the paper https://arxiv.org/pdf/1909.13144.pdf.

### Test Plan:
Run unit tests with: `python pytorch/test/quantization/core/experimental/test_quantized_tensor.py`

[ghstack-poisoned]

asl3 committed:
Update on "[quant] Implement quantize APoT method"


### Summary:
This PR implements functionality to quantize input floating point tensors. The method takes a floating point tensor as input and converts to APoT integer representation based on the calculated quantization levels from the specified APoT non-uniform observer. This approach follows the methodology outlined in the paper https://arxiv.org/pdf/1909.13144.pdf.

### Test Plan:
Run unit tests with: `python pytorch/test/quantization/core/experimental/test_quantized_tensor.py`

[ghstack-poisoned]

asl3 committed:
Update on "[quant] Implement quantize APoT method"


### Summary:
This PR implements functionality to quantize input floating point tensors. The method takes a floating point tensor as input and converts to APoT integer representation based on the calculated quantization levels from the specified APoT non-uniform observer. This approach follows the methodology outlined in the paper https://arxiv.org/pdf/1909.13144.pdf.

### Test Plan:
Run unit tests with: `python pytorch/test/quantization/core/experimental/test_quantized_tensor.py`

[ghstack-poisoned]

dzdang commented on test/quantization/core/experimental/test_quantized_tensor.py:
```diff
@@ -1,19 +1,116 @@
-        with self.assertRaises(NotImplementedError):
-            TensorAPoT.quantize_APoT(t)
+    r""" Tests quantize_APoT result (int representation) on random 1-dim tensor
+        and hardcoded values for b, k by comparing to uniform observer
```
seems this was not addressed? you're not comparing to uniform "observer". you're comparing to uniform quantzation

dzdang commented on test/quantization/core/experimental/test_quantized_tensor.py:
```diff
@@ -1,19 +1,116 @@
-            TensorAPoT.quantize_APoT(t)
+    r""" Tests quantize_APoT result (int representation) on random 1-dim tensor
+        and hardcoded values for b, k by comparing to uniform observer
+        (non-uniform observer reduces to uniform for k = 1)
```
same here

dzdang commented on test/quantization/core/experimental/test_quantized_tensor.py:
```diff
@@ -1,19 +1,113 @@
+        uniform_quantized = quantize_per_tensor(input=tensor2quantize, scale=1.0, zero_point=0, dtype=torch.quint8).int_repr()
+
+        qtensor_data = torch.tensor(qtensor.data).type(torch.uint8)
+        uniform_quantized_tens = uniform_quantized.data
```
let's just avoid using "tens" to mean "tensor"

asl3 committed:
Update on "[quant] Implement quantize APoT method"


### Summary:
This PR implements functionality to quantize input floating point tensors. The method takes a floating point tensor as input and converts to APoT integer representation based on the calculated quantization levels from the specified APoT non-uniform observer. This approach follows the methodology outlined in the paper https://arxiv.org/pdf/1909.13144.pdf.

### Test Plan:
Run unit tests with: `python pytorch/test/quantization/core/experimental/test_quantized_tensor.py`

[ghstack-poisoned]

jerryzh168 commented on torch/ao/quantization/experimental/APoT_tensor.py:
```diff
@@ -1,13 +1,62 @@
-    @staticmethod
-    def quantize_APoT(tensor2quantize: Tensor) -> Tensor:
-        raise NotImplementedError
+    b: int
```
this is not exactly following the C++ Tensor design actually, wondering if we want to fix it.
What we have in C++ is:
```
Quantizer class that has quantize and dequantize method and stores all the quantization parameters: https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/quantized/Quantizer.h#L78

Quantized Tensor class that holds a quantizer object, stores integer representation in data, dtype will be quint8/qint8 etc.
int_repr is also a method on the quantized tensor, that will convert the quint8/qint8 tensor to uint8/int8 tensor and return the result
```

asl3 replied:
yes, will fix

asl3 committed:
Update on "[quant] Implement quantize APoT method"


### Summary:
This PR implements functionality to quantize input floating point tensors. The method takes a floating point tensor as input and converts to APoT integer representation based on the calculated quantization levels from the specified APoT non-uniform observer. This approach follows the methodology outlined in the paper https://arxiv.org/pdf/1909.13144.pdf.

### Test Plan:
Run unit tests with: `python pytorch/test/quantization/core/experimental/test_quantized_tensor.py`

[ghstack-poisoned]

asl3 committed:
Update on "[quant] Implement quantize APoT method"


### Summary:
This PR implements functionality to quantize input floating point tensors. The method takes a floating point tensor as input and converts to APoT integer representation based on the calculated quantization levels from the specified APoT non-uniform observer. This approach follows the methodology outlined in the paper https://arxiv.org/pdf/1909.13144.pdf.

### Test Plan:
Run unit tests with: `python pytorch/test/quantization/core/experimental/test_quantized_tensor.py`

[ghstack-poisoned]

asl3 committed:
Update on "[quant] Implement quantize APoT method"


### Summary:
This PR implements functionality to quantize input floating point tensors. The method takes a floating point tensor as input and converts to APoT integer representation based on the calculated quantization levels from the specified APoT non-uniform observer. This approach follows the methodology outlined in the paper https://arxiv.org/pdf/1909.13144.pdf.

### Test Plan:
Run unit tests with: `python pytorch/test/quantization/core/experimental/test_quantized_tensor.py`

[ghstack-poisoned]

asl3 committed:
Update on "[quant] Implement quantize APoT method"


### Summary:
This PR implements functionality to quantize input floating point tensors. The method takes a floating point tensor as input and converts to APoT integer representation based on the calculated quantization levels from the specified APoT non-uniform observer. This approach follows the methodology outlined in the paper https://arxiv.org/pdf/1909.13144.pdf.

### Test Plan:
Run unit tests with: `python pytorch/test/quantization/core/experimental/test_quantized_tensor.py`

[ghstack-poisoned]

asl3 committed:
Update on "[quant] Implement quantize APoT method"


### Summary:
This PR implements functionality to quantize input floating point tensors. The method takes a floating point tensor as input and converts to APoT integer representation based on the calculated quantization levels from the specified APoT non-uniform observer. This approach follows the methodology outlined in the paper https://arxiv.org/pdf/1909.13144.pdf.

### Test Plan:
Run unit tests with: `python pytorch/test/quantization/core/experimental/test_quantized_tensor.py`

[ghstack-poisoned]

asl3 committed:
Update on "[quant] Implement quantize APoT method"


### Summary:
This PR implements functionality to quantize input floating point tensors. The method takes a floating point tensor as input and converts to APoT integer representation based on the calculated quantization levels from the specified APoT non-uniform observer. This approach follows the methodology outlined in the paper https://arxiv.org/pdf/1909.13144.pdf.

### Test Plan:
Run unit tests with: `python pytorch/test/quantization/core/experimental/test_quantized_tensor.py`

[ghstack-poisoned]

asl3 committed:
Update on "[quant] Implement quantize APoT method"


### Summary:
This PR implements functionality to quantize input floating point tensors. The method takes a floating point tensor as input and converts to APoT integer representation based on the calculated quantization levels from the specified APoT non-uniform observer. This approach follows the methodology outlined in the paper https://arxiv.org/pdf/1909.13144.pdf.

### Test Plan:
Run unit tests with: `python pytorch/test/quantization/core/experimental/test_quantized_tensor.py`

[ghstack-poisoned]

asl3 committed:
Update on "[quant] Implement quantize APoT method"


### Summary:
This PR implements functionality to quantize input floating point tensors. The method takes a floating point tensor as input and converts to APoT integer representation based on the calculated quantization levels from the specified APoT non-uniform observer. This approach follows the methodology outlined in the paper https://arxiv.org/pdf/1909.13144.pdf.

### Test Plan:
Run unit tests with: `python pytorch/test/quantization/core/experimental/test_quantized_tensor.py`

[ghstack-poisoned]

asl3 committed:
Update on "[quant] Implement quantize APoT method"


### Summary:
This PR implements functionality to quantize input floating point tensors. The method takes a floating point tensor as input and converts to APoT integer representation based on the calculated quantization levels from the specified APoT non-uniform observer. This approach follows the methodology outlined in the paper https://arxiv.org/pdf/1909.13144.pdf.

### Test Plan:
Run unit tests with: `python pytorch/test/quantization/core/experimental/test_quantized_tensor.py`

[ghstack-poisoned]

jerryzh168 commented on test/quantization/core/experimental/test_quantizer.py:
```diff
@@ -1,19 +1,144 @@
+        expectedResult = True
+
+        for result, orig in zip(qtensor_data, tensor2quantize):
+            if abs(result - orig) > 0.25:
```
there is no such guarantee actually, I think what we do typically is have a reference implementation in python and compare with that

asl3 replied:
decided to instead test using hand-calculated example from APoT paper (https://arxiv.org/pdf/1909.13144.pdf) after talking with @dzdang

jerryzh168 commented on torch/ao/quantization/experimental/quantizer.py:
```diff
@@ -1,13 +1,64 @@
+    The approach follows the method outlined in the APoT paper: https://arxiv.org/pdf/1909.13144.pdf.
+    Args:
+        tensor2quantize: fp Tensor
+        use_int_repr: bool flag to specify int of reduced precision fp representation of APoT tensor
```
what is the difference between True and False? can you give an example?

jerryzh168 replied:
looks like reduced_precision is the dequantized fp value. I feel we should only be storing int values in `data` so that representation is in integer

dzdang replied:
@jerryzh168 I think reduced_precision is the quantized levels in fp representation (powers of 2) so I think it's analogous to dequantized fp values in uniform quantzation?

jerryzh168 replied:
yeah, but we don't really store dequantized fp values in uniform quantization, the reason we quantize things is so that we can store things in int8 I think. I'm wondering why do we store this?

asl3 replied:
i'll modify so that we don't store reduced precision representation anymore (just int8 representation)

asl3 committed:
Update on "[quant] Implement quantize APoT method"


### Summary:
This PR implements functionality to quantize input floating point tensors. The method takes a floating point tensor as input and converts to APoT integer representation based on the calculated quantization levels from the specified APoT non-uniform observer. This approach follows the methodology outlined in the paper https://arxiv.org/pdf/1909.13144.pdf.

### Test Plan:
Run unit tests with: `python pytorch/test/quantization/core/experimental/test_quantizer.py`

[ghstack-poisoned]

asl3 committed:
Update on "[quant] Implement quantize APoT method"


### Summary:
This PR implements functionality to quantize input floating point tensors. The method takes a floating point tensor as input and converts to APoT integer representation based on the calculated quantization levels from the specified APoT non-uniform observer. This approach follows the methodology outlined in the paper https://arxiv.org/pdf/1909.13144.pdf.

### Test Plan:
Run unit tests with: `python pytorch/test/quantization/core/experimental/test_quantizer.py`

[ghstack-poisoned]

dzdang reviewed (approved):
looks good to me, but what is the purpose of having both a 1d and 2d test? is there a dependency on the number of dimensions in your implementation?

asl3 committed:
Update on "[quant] Implement quantize APoT method"


### Summary:
This PR implements functionality to quantize input floating point tensors. The method takes a floating point tensor as input and converts to APoT integer representation based on the calculated quantization levels from the specified APoT non-uniform observer. This approach follows the methodology outlined in the paper https://arxiv.org/pdf/1909.13144.pdf.

### Test Plan:
Run unit tests with: `python pytorch/test/quantization/core/experimental/test_quantizer.py`

[ghstack-poisoned]

asl3 committed:
Update on "[quant] Implement quantize APoT method"


### Summary:
This PR implements functionality to quantize input floating point tensors. The method takes a floating point tensor as input and converts to APoT integer representation based on the calculated quantization levels from the specified APoT non-uniform observer. This approach follows the methodology outlined in the paper https://arxiv.org/pdf/1909.13144.pdf.

### Test Plan:
Run unit tests with: `python pytorch/test/quantization/core/experimental/test_quantizer.py`

[ghstack-poisoned]

asl3 committed:
Update on "[quant] Implement quantize APoT method"


### Summary:
This PR implements functionality to quantize input floating point tensors. The method takes a floating point tensor as input and converts to APoT integer representation based on the calculated quantization levels from the specified APoT non-uniform observer. This approach follows the methodology outlined in the paper https://arxiv.org/pdf/1909.13144.pdf.

### Test Plan:
Run unit tests with: `python pytorch/test/quantization/core/experimental/test_quantizer.py`

[ghstack-poisoned]

asl3 committed:
Update on "[quant] Implement quantize APoT method"


### Summary:
This PR implements functionality to quantize input floating point tensors. The method takes a floating point tensor as input and converts to APoT integer representation based on the calculated quantization levels from the specified APoT non-uniform observer. This approach follows the methodology outlined in the paper https://arxiv.org/pdf/1909.13144.pdf.

### Test Plan:
Run unit tests with: `python pytorch/test/quantization/core/experimental/test_quantizer.py`

[ghstack-poisoned]

asl3 committed:
Update on "[quant] Implement quantize APoT method"


### Summary:
This PR implements functionality to quantize input floating point tensors. The method takes a floating point tensor as input and converts to APoT integer representation based on the calculated quantization levels from the specified APoT non-uniform observer. This approach follows the methodology outlined in the paper https://arxiv.org/pdf/1909.13144.pdf.

### Test Plan:
Run unit tests with: `python pytorch/test/quantization/core/experimental/test_quantizer.py`

[ghstack-poisoned]

asl3 committed:
Update on "[quant] Implement quantize APoT method"


### Summary:
This PR implements functionality to quantize input floating point tensors. The method takes a floating point tensor as input and converts to APoT integer representation based on the calculated quantization levels from the specified APoT non-uniform observer. This approach follows the methodology outlined in the paper https://arxiv.org/pdf/1909.13144.pdf.

### Test Plan:
Run unit tests with: `python pytorch/test/quantization/core/experimental/test_quantizer.py`

[ghstack-poisoned]

asl3 commented:
@pytorchbot merge -g

asl3 committed:
Update on "[quant] Implement quantize APoT method"


### Summary:
This PR implements functionality to quantize input floating point tensors. The method takes a floating point tensor as input and converts to APoT integer representation based on the calculated quantization levels from the specified APoT non-uniform observer. This approach follows the methodology outlined in the paper https://arxiv.org/pdf/1909.13144.pdf.

### Test Plan:
Run unit tests with: `python pytorch/test/quantization/core/experimental/test_quantizer.py`

[ghstack-poisoned]

asl3 committed:
Update on "[quant] Implement quantize APoT method"


### Summary:
This PR implements functionality to quantize input floating point tensors. The method takes a floating point tensor as input and converts to APoT integer representation based on the calculated quantization levels from the specified APoT non-uniform observer. This approach follows the methodology outlined in the paper https://arxiv.org/pdf/1909.13144.pdf.

### Test Plan:
Run unit tests with: `python pytorch/test/quantization/core/experimental/test_quantizer.py`

[ghstack-poisoned]

asl3 committed:
Update on "[quant] Implement quantize APoT method"


### Summary:
This PR implements functionality to quantize input floating point tensors. The method takes a floating point tensor as input and converts to APoT integer representation based on the calculated quantization levels from the specified APoT non-uniform observer. This approach follows the methodology outlined in the paper https://arxiv.org/pdf/1909.13144.pdf.

### Test Plan:
Run unit tests with: `python pytorch/test/quantization/core/experimental/test_quantizer.py`

[ghstack-poisoned]

jerryzh168 commented on test/quantization/core/experimental/test_quantizer.py:
```diff
@@ -1,14 +1,81 @@
 from torch.ao.quantization.experimental.quantizer import APoTQuantizer
 import unittest
+import random
+quantize_APoT = APoTQuantizer.quantize_APoT
+dequantize = APoTQuantizer.dequantize
```
are these used?

jerryzh168 reviewed (approved):
LGTM, thanks for addressing the comments

jerryzh168 commented on torch/ao/quantization/experimental/quantizer.py:
```diff
@@ -1,12 +1,58 @@
+    signed: bool
+    quantization_levels: torch.Tensor
+    level_indices: torch.Tensor
+    data: torch.Tensor
```
if we want to be aligned with C++ api, data should probably be stored in a quantized tensor I think, quantizer can store all the quantization parameters

dzdang replied:
I think we should make this change. missed this earlier. data isn't an attribute of the quantizer

jerryzh168 commented on torch/ao/quantization/experimental/quantizer.py:
```diff
@@ -1,12 +1,58 @@
+        self.n = b // k
+
+        # make observer, get quantizion levels and level indices
+        obs = APoTObserver(max_val=max_val, b=b, k=k)
+        obs_result = obs.calculate_qparams(signed=signed)
```
I feel this observer is not used the same way as the other observers in the code base, since it does not observe anything, is `max_val` something we get from observed Tensor? would it make more sense to move this argument to `calculate_qparams`? meaning that we are overriding the `max_val` instead from the internal state of observer while calculating quantization paramters.

asl3 replied:
yes good point, I can make a new PR to modify so calculate_qparams is the one that takes in max_val

asl3 committed:
Update on "[quant] Implement quantize APoT method"


### Summary:
This PR implements functionality to quantize input floating point tensors. The method takes a floating point tensor as input and converts to APoT integer representation based on the calculated quantization levels from the specified APoT non-uniform observer. This approach follows the methodology outlined in the paper https://arxiv.org/pdf/1909.13144.pdf.

### Test Plan:
Run unit tests with: `python pytorch/test/quantization/core/experimental/test_quantizer.py`

[ghstack-poisoned]

asl3 commented:
@pytorchbot merge -g
